### ISSUE TITLE
[BUG] Ocasional crashes in av. offline folders with many items

### ISSUE THREAD

**Issue by jesmrec** [owncloud/android#2574]
There are not specific steps to reproduce, i will detail the process i did:

1. I have a folder called "Documents" in root. It contains 250 small txt files (~1 KB each)
2. Set the "Documents" folder as available offline. Everything inside the folder is dowloaded and marked as av. offline

From this point, if you browse inside the folder and then browse back to root, the folder is synced (it is av. offline, therefore always synced). By repeating this process several times (no more than 3 - 4), the app finally crashes

I reproduced it with Android 9 and 8. Will update if i reproduce with 7 or older.

I have noticed two different stack traces:

```
2019-06-12 09:52:33.008 26366-26434/? E/AndroidRuntime: FATAL EXCEPTION: Thread-6
    Process: com.owncloud.android.debug, PID: 26366
    java.lang.ArrayIndexOutOfBoundsException: length=10; index=-1
        at java.util.ArrayList.add(ArrayList.java:468)
        at com.owncloud.android.lib.common.http.interceptors.HttpInterceptor.addRequestInterceptor(HttpInterceptor.java:63)
        at com.owncloud.android.lib.common.http.HttpClient.addHeaderForAllRequests(HttpClient.java:160)
        at com.owncloud.android.lib.common.authentication.OwnCloudBearerCredentials.applyTo(OwnCloudBearerCredentials.java:46)
        at com.owncloud.android.lib.common.OwnCloudClient.applyCredentials(OwnCloudClient.java:103)
        at com.owncloud.android.lib.common.SingleSessionManager.keepCredentialsUpdated(SingleSessionManager.java:213)
        at com.owncloud.android.lib.common.SingleSessionManager.getClientFor(SingleSessionManager.java:143)
        at com.owncloud.android.lib.common.DynamicSessionManager.getClientFor(DynamicSessionManager.java:40)
        at com.owncloud.android.lib.common.operations.RemoteOperation.grantOwnCloudClient(RemoteOperation.java:143)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:250)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:190)
        at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:68)
        at com.owncloud.android.operations.SynchronizeFolderOperation.syncContents(SynchronizeFolderOperation.java:484)
        at com.owncloud.android.operations.SynchronizeFolderOperation.run(SynchronizeFolderOperation.java:227)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:251)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:213)
        at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:87)
        at com.owncloud.android.operations.RefreshFolderOperation.run(RefreshFolderOperation.java:142)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:251)
        at com.owncloud.android.lib.common.operations.RemoteOperation.run(RemoteOperation.java:270)
        at java.lang.Thread.run(Thread.java:764)
2019-06-12 09:52:33.008 26366-26391/? E/AndroidRuntime: FATAL EXCEPTION: Syncfolder thread
    Process: com.owncloud.android.debug, PID: 26366
    java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.owncloud.android.lib.common.operations.RemoteOperationResult.isSuccess()' on a null object reference
        at com.owncloud.android.services.SyncFolderHandler.doOperation(SyncFolderHandler.java:136)
        at com.owncloud.android.services.SyncFolderHandler.handleMessage(SyncFolderHandler.java:94)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
2019-06-12 09:52:33.048 1173-1360/? W/InputDispatcher: channel '461abfe com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
2019-06-12 09:52:33.048 1173-1360/? E/InputDispatcher: channel '461abfe com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
2019-06-12 09:52:33.048 1173-1677/? I/WindowManager: WIN DEATH: Window{461abfe u0 com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity}
2019-06-12 09:52:33.048 1173-1677/? W/InputDispatcher: Attempted to unregister already unregistered input channel '461abfe com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)'
2019-06-12 09:52:33.050 1173-1194/? I/ActivityManager: Process com.owncloud.android.debug (pid 26366) has died: fore TOP 
2019-06-12 09:52:33.051 1173-1194/? W/ActivityManager: Scheduling restart of crashed service com.owncloud.android.debug/com.owncloud.android.services.OperationsService in 1000ms
2019-06-12 09:52:33.053 1173-1194/? W/ActivityManager: Force removing ActivityRecord{d234752 u0 com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity t661}: app died, no saved state
2019-06-12 09:52:33.055 606-730/? W/SurfaceFlinger: Attempting to destroy on removed layer: AppWindowToken{61fab20 token=Token{32c8a23 ActivityRecord{d234752 u0 com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity t661}}}#0
```

```
2019-06-12 09:53:40.579 26451-26489/com.owncloud.android.debug E/AndroidRuntime: FATAL EXCEPTION: Thread-3
    Process: com.owncloud.android.debug, PID: 26451
    java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.next(ArrayList.java:860)
        at com.owncloud.android.lib.common.http.interceptors.HttpInterceptor.deleteRequestHeaderInterceptor(HttpInterceptor.java:91)
        at com.owncloud.android.lib.common.http.HttpClient.deleteHeaderForAllRequests(HttpClient.java:166)
        at com.owncloud.android.lib.common.authentication.OwnCloudBearerCredentials.applyTo(OwnCloudBearerCredentials.java:43)
        at com.owncloud.android.lib.common.OwnCloudClient.applyCredentials(OwnCloudClient.java:103)
        at com.owncloud.android.lib.common.SingleSessionManager.keepCredentialsUpdated(SingleSessionManager.java:213)
        at com.owncloud.android.lib.common.SingleSessionManager.getClientFor(SingleSessionManager.java:143)
        at com.owncloud.android.lib.common.DynamicSessionManager.getClientFor(DynamicSessionManager.java:40)
        at com.owncloud.android.lib.common.operations.RemoteOperation.grantOwnCloudClient(RemoteOperation.java:143)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:250)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:190)
        at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:68)
        at com.owncloud.android.operations.SynchronizeFolderOperation.syncContents(SynchronizeFolderOperation.java:484)
        at com.owncloud.android.operations.SynchronizeFolderOperation.run(SynchronizeFolderOperation.java:227)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:251)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:213)
        at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:87)
        at com.owncloud.android.operations.RefreshFolderOperation.run(RefreshFolderOperation.java:142)
        at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:251)
        at com.owncloud.android.lib.common.operations.RemoteOperation.run(RemoteOperation.java:270)
        at java.lang.Thread.run(Thread.java:764)
2019-06-12 09:53:40.582 1173-1194/? W/ActivityManager:   Force finishing activity com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity
2019-06-12 09:53:40.626 1173-1360/? W/InputDispatcher: channel '18801fb com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
2019-06-12 09:53:40.626 1173-1360/? E/InputDispatcher: channel '18801fb com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
2019-06-12 09:53:40.627 1173-8290/? I/WindowManager: WIN DEATH: Window{18801fb u0 com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity}
2019-06-12 09:53:40.627 1173-8290/? W/InputDispatcher: Attempted to unregister already unregistered input channel '18801fb com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity (server)'
2019-06-12 09:53:40.627 1173-6237/? I/ActivityManager: Process com.owncloud.android.debug (pid 26451) has died: vis  +99TOP 
2019-06-12 09:53:40.627 1173-6237/? W/ActivityManager: Scheduling restart of crashed service com.owncloud.android.debug/com.owncloud.android.services.OperationsService in 1000ms
2019-06-12 09:53:40.631 606-1349/? W/SurfaceFlinger: Attempting to set client state on removed layer: com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity#0
2019-06-12 09:53:40.631 606-1349/? W/SurfaceFlinger: Attempting to destroy on removed layer: com.owncloud.android.debug/com.owncloud.android.ui.activity.FileDisplayActivity#0
```

**Comment by abelgardep** [comment]
I was able to reproduce the second stacktrace.

Caused by concurrency on the HttpInterceptor.java `deleteRequestHeaderInterceptor(String headerName)`

Pixel 2 v10

**Comment by davigonz** [comment]
I see here two errors:

 - [x] `java.lang.NullPointerException: Attempt to invoke virtual method 'boolean com.owncloud.android.lib.common.operations.RemoteOperationResult.isSuccess()' on a null object reference at com.owncloud.android.services.SyncFolderHandler.doOperation(SyncFolderHandler.java:136)` => already fixed via https://github.com/owncloud/android/pull/2796

 - [ ] `java.util.ConcurrentModificationException`

**Comment by abelgardep** [comment]
To fix this issue, we will introduce headers in OwnCloudClient and not in interceptor.

**Comment by jesmrec** [comment]
Fixed in 2.15.1
